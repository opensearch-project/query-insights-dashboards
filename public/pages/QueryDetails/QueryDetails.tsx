/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useCallback, useEffect } from 'react';
import Plotly from 'plotly.js-dist';
import {
  EuiButton,
  EuiCodeBlock,
  EuiFlexGrid,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiPanel,
  EuiSpacer,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import hash from 'object-hash';
import { useParams, useHistory } from 'react-router-dom';
import { CoreStart } from 'opensearch-dashboards/public';
import QuerySummary from './Components/QuerySummary';
import { QUERY_INSIGHTS } from '../TopNQueries/TopNQueries';
import { SearchQueryRecord } from '../../../types/types';
import { PageHeader } from '../../components/PageHeader';
import { QueryInsightsDashboardsPluginStartDependencies } from '../../types';

const QueryDetails = ({
  queries,
  core,
  depsStart,
}: {
  queries: any;
  core: CoreStart;
  depsStart: QueryInsightsDashboardsPluginStartDependencies;
}) => {
  const { hashedQuery } = useParams<{ hashedQuery: string }>();
  const query = queries.find((q: SearchQueryRecord) => hash(q) === hashedQuery);
  const history = useHistory();

  // Convert UNIX time to a readable format
  const convertTime = useCallback((unixTime: number) => {
    const date = new Date(unixTime);
    const [_weekDay, month, day, year] = date.toDateString().split(' ');
    return `${month} ${day}, ${year} @ ${date.toLocaleTimeString('en-US')}`;
  }, []);

  // Initialize the Plotly chart
  const initPlotlyChart = useCallback(() => {
    const latencies: number[] = Object.values(query?.phase_latency_map || [0, 0, 0]);
    const data = [
      {
        x: latencies.reverse(),
        y: ['Fetch    ', 'Query    ', 'Expand    '],
        type: 'bar',
        orientation: 'h',
        width: 0.5,
        marker: { color: ['#F990C0', '#1BA9F5', '#7DE2D1'] },
        base: [latencies[2] + latencies[1], latencies[2], 0],
        text: latencies.map((value) => `${value}ms`),
        textposition: 'outside',
        cliponaxis: false,
      },
    ];
    const layout = {
      autosize: true,
      margin: { l: 80, r: 80, t: 25, b: 15, pad: 0 },
      height: 120,
      xaxis: {
        side: 'top',
        zeroline: false,
        ticksuffix: 'ms',
        autorangeoptions: { clipmin: 0 },
        tickfont: { color: '#535966' },
        linecolor: '#D4DAE5',
        gridcolor: '#D4DAE5',
      },
      yaxis: { linecolor: '#D4DAE5' },
    };
    const config = { responsive: true };
    Plotly.newPlot('latency', data, layout, config);
  }, [query]);

  useEffect(() => {
    core.chrome.setBreadcrumbs([
      {
        text: 'Query insights',
        href: QUERY_INSIGHTS,
        onClick: (e) => {
          e.preventDefault();
          history.push(QUERY_INSIGHTS);
        },
      },
      { text: `Query details: ${convertTime(query.timestamp)}` },
    ]);
    initPlotlyChart();
  }, [query, history, core.chrome, convertTime, initPlotlyChart]);

  const queryString = JSON.stringify(query.source, null, 2);
  const queryDisplay = `{\n  "query": ${queryString ? queryString.replace(/\n/g, '\n  ') : ''}\n}`;

  return (
    <div>
      <PageHeader
        coreStart={core}
        depsStart={depsStart}
        fallBackComponent={
          <>
            <EuiTitle size="l">
              <h1>Query details</h1>
            </EuiTitle>
            <EuiSpacer size="l" />
          </>
        }
      />

      <EuiFlexItem>
        <QuerySummary query={query} />
        <EuiSpacer size="m" />
        <EuiFlexGrid columns={2}>
          <EuiFlexItem grow={1}>
            <EuiPanel data-test-subj={'query-details-source-section'}>
              <EuiFlexGroup alignItems="center" justifyContent="spaceBetween">
                <EuiFlexItem>
                  <EuiText size="xs">
                    <h2>Query</h2>
                  </EuiText>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiButton
                    iconSide="right"
                    iconType="popout"
                    target="_blank"
                    href="https://playground.opensearch.org/app/searchRelevance#/"
                  >
                    Open in search comparison
                  </EuiButton>
                </EuiFlexItem>
              </EuiFlexGroup>
              <EuiHorizontalRule margin="xs" />
              <EuiSpacer size="xs" />
              <EuiCodeBlock
                language="jsx"
                paddingSize="m"
                fontSize="s"
                overflowHeight={600}
                isCopyable
              >
                {queryDisplay}
              </EuiCodeBlock>
            </EuiPanel>
          </EuiFlexItem>
          <EuiFlexItem grow={1} style={{ alignSelf: 'start' }}>
            <EuiPanel data-test-subj={'query-details-latency-chart'}>
              <EuiText size="xs">
                <h2>Latency</h2>
              </EuiText>
              <EuiHorizontalRule margin="m" />
              <div id="latency" />
            </EuiPanel>
          </EuiFlexItem>
        </EuiFlexGrid>
      </EuiFlexItem>
    </div>
  );
};

// eslint-disable-next-line import/no-default-export
export default QueryDetails;
