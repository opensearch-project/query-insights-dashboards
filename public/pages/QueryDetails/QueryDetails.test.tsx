/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { render, screen } from '@testing-library/react';
import QueryDetails from './QueryDetails';
import Plotly from 'plotly.js-dist';
import { MockQueries } from '../../../test/testUtils';
import '@testing-library/jest-dom';
import { MemoryRouter, Route } from 'react-router-dom';
import hash from 'object-hash';
// Mock the external dependencies
jest.mock('plotly.js-dist', () => ({
  newPlot: jest.fn(),
}));

const mockCoreStart = {
  chrome: {
    setBreadcrumbs: jest.fn(),
  },
};
const mockQuery = MockQueries()[0];
const mockParams = { hashedQuery: hash(mockQuery) };

describe('QueryDetails component', () => {
  beforeAll(() => {
    jest.spyOn(Date.prototype, 'toLocaleTimeString').mockImplementation(() => '12:00:00 AM');
    jest.spyOn(Date.prototype, 'toDateString').mockImplementation(() => 'Mon Jan 13 2025');
  });

  afterAll(() => {
    jest.resetAllMocks(); // Reset all mocks after all tests
  });

  beforeEach(() => {
    jest.clearAllMocks(); // Clear all mock calls and instances before each test
  });

  const renderComponent = () =>
    render(
      <MemoryRouter initialEntries={[`/query-details/${mockParams.hashedQuery}`]}>
        <Route path="/query-details/:hashedQuery">
          <QueryDetails queries={MockQueries()} core={mockCoreStart} />
        </Route>
      </MemoryRouter>
    );

  it('renders the QueryDetails page', () => {
    const { container } = renderComponent();
    // Check if the query details are displayed correctly
    expect(screen.getByText('Query details')).toBeInTheDocument();
    expect(screen.getByText('Query')).toBeInTheDocument();

    // Verify that the Plotly chart is rendered
    expect(Plotly.newPlot).toHaveBeenCalledTimes(1);
    // Verify the breadcrumbs were set correctly
    expect(mockCoreStart.chrome.setBreadcrumbs).toHaveBeenCalled();

    expect(container).toMatchSnapshot();
  });
});
