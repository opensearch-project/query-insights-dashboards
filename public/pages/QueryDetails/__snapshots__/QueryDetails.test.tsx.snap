// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryDetails component renders the QueryDetails page 1`] = `
<div>
  <div>
    <h1
      class="euiTitle euiTitle--large"
    >
      Query details
    </h1>
    <div
      class="euiSpacer euiSpacer--l"
    />
    <div
      class="euiFlexItem"
    >
      <div
        class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
        data-test-subj="query-details-summary-section"
      >
        <div
          class="euiText euiText--extraSmall"
        >
          <h2>
            Summary
          </h2>
        </div>
        <hr
          class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginMedium"
        />
        <div
          class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--fourths euiFlexGrid--responsive"
        >
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Timestamp
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              Jan 13, 2025 @ 12:00:00 AM
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Latency
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              8 ms
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                CPU Time
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              5.51 ms
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Memory Usage
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              N/A B
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Indices
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              top_queries-2024.09.12
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Search Type
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              query then fetch
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Coordinator Node ID
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              Q36D2z_NRGKim6EZZMgi6A
            </div>
          </div>
          <div
            class="euiFlexItem"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h4>
                Total Shards
              </h4>
            </div>
            <div
              class="euiText euiText--extraSmall"
            >
              1
            </div>
          </div>
        </div>
      </div>
      <div
        class="euiSpacer euiSpacer--m"
      />
      <div
        class="euiFlexGrid euiFlexGrid--gutterLarge euiFlexGrid--halves euiFlexGrid--responsive"
      >
        <div
          class="euiFlexItem euiFlexItem--flexGrow1"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
            data-test-subj="query-details-source-section"
          >
            <div
              class="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--justifyContentSpaceBetween euiFlexGroup--directionRow euiFlexGroup--responsive"
            >
              <div
                class="euiFlexItem"
              >
                <div
                  class="euiText euiText--extraSmall"
                >
                  <h2>
                    Query
                  </h2>
                </div>
              </div>
              <div
                class="euiFlexItem euiFlexItem--flexGrowZero"
              >
                <a
                  class="euiButton euiButton--primary"
                  href="https://playground.opensearch.org/app/searchRelevance#/"
                  rel="noopener noreferrer"
                  target="_blank"
                >
                  <span
                    class="euiButtonContent euiButtonContent--iconRight euiButton__content"
                  >
                    <svg
                      aria-hidden="true"
                      class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonContent__icon"
                      focusable="false"
                      height="16"
                      role="img"
                      viewBox="0 0 16 16"
                      width="16"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                      />
                    </svg>
                    <span
                      class="euiButton__text"
                    >
                      Open in search comparison
                    </span>
                  </span>
                </a>
              </div>
            </div>
            <hr
              class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginXSmall"
            />
            <div
              class="euiSpacer euiSpacer--xs"
            />
            <div
              class="euiCodeBlock euiCodeBlock--fontSmall euiCodeBlock--paddingMedium euiCodeBlock--hasControls prismjs language-jsx"
              style="max-height: 600px;"
            >
              <pre
                class="euiCodeBlock__pre euiCodeBlock__pre--whiteSpacePreWrap"
                style="max-height: 600px;"
                tabindex="-1"
              >
                <code
                  class="euiCodeBlock__code jsx"
                >
                  <span
                    class="euiCodeBlock__line"
                  >
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                      
                    <span
                      class="token string-property property"
                    >
                      "query"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                        
                    <span
                      class="token string-property property"
                    >
                      "query"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                          
                    <span
                      class="token string-property property"
                    >
                      "bool"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                            
                    <span
                      class="token string-property property"
                    >
                      "must"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      [
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                              
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                
                    <span
                      class="token string-property property"
                    >
                      "range"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                  
                    <span
                      class="token string-property property"
                    >
                      "timestamp"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "from"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token number"
                    >
                      1726092595177
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "to"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token number"
                    >
                      1726178995177
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "include_lower"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token boolean"
                    >
                      true
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "include_upper"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token boolean"
                    >
                      true
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "boost"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token number"
                    >
                      1
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                  
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                              
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                            
                    <span
                      class="token punctuation"
                    >
                      ]
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                            
                    <span
                      class="token string-property property"
                    >
                      "must_not"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      [
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                              
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                
                    <span
                      class="token string-property property"
                    >
                      "match"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                  
                    <span
                      class="token string-property property"
                    >
                      "indices"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token punctuation"
                    >
                      {
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "query"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token string"
                    >
                      "top_queries*"
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "operator"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token string"
                    >
                      "OR"
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "prefix_length"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token number"
                    >
                      0
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "max_expansions"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token number"
                    >
                      50
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "fuzzy_transpositions"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token boolean"
                    >
                      true
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "lenient"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token boolean"
                    >
                      false
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "zero_terms_query"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token string"
                    >
                      "NONE"
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "auto_generate_synonyms_phrase_query"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token boolean"
                    >
                      true
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                    
                    <span
                      class="token string-property property"
                    >
                      "boost"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token number"
                    >
                      1
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                  
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                                
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                              
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                            
                    <span
                      class="token punctuation"
                    >
                      ]
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                            
                    <span
                      class="token string-property property"
                    >
                      "adjust_pure_negative"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token boolean"
                    >
                      true
                    </span>
                    <span
                      class="token punctuation"
                    >
                      ,
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                            
                    <span
                      class="token string-property property"
                    >
                      "boost"
                    </span>
                    <span
                      class="token operator"
                    >
                      :
                    </span>
                     
                    <span
                      class="token number"
                    >
                      1
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                          
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                        
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                      
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                    

                  </span>
                  <span
                    class="euiCodeBlock__line"
                  >
                    
                    <span
                      class="token punctuation"
                    >
                      }
                    </span>
                  </span>
                </code>
              </pre>
              <div
                class="euiCodeBlock__controls"
              >
                <button
                  aria-label="Expand"
                  class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall euiCodeBlock__fullScreenButton"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                    focusable="false"
                    height="16"
                    role="img"
                    viewBox="0 0 16 16"
                    width="16"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                    />
                  </svg>
                </button>
                <div
                  class="euiCodeBlock__copyButton"
                >
                  <span
                    class="euiToolTipAnchor"
                  >
                    <button
                      aria-label="Copy"
                      class="euiButtonIcon euiButtonIcon--text euiButtonIcon--empty euiButtonIcon--xSmall"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiIcon--inherit euiIcon-isLoading euiButtonIcon__icon"
                        focusable="false"
                        height="16"
                        role="img"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M5.277 10.088c.02.014.04.03.057.047.582.55 1.134.812 1.666.812.586 0 1.84-.293 3.713-.88L9 6.212V2H7v4.212l-1.723 3.876Zm-.438.987L3.539 14h8.922l-1.32-2.969C9.096 11.677 7.733 12 7 12c-.74 0-1.463-.315-2.161-.925ZM6 2H5V1h6v1h-1v4l3.375 7.594A1 1 0 0 1 12.461 15H3.54a1 1 0 0 1-.914-1.406L6 6V2Z"
                        />
                      </svg>
                    </button>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiFlexItem euiFlexItem--flexGrow1"
          style="align-self: start;"
        >
          <div
            class="euiPanel euiPanel--paddingMedium euiPanel--borderRadiusMedium euiPanel--plain euiPanel--hasShadow"
            data-test-subj="query-details-latency-chart"
          >
            <div
              class="euiText euiText--extraSmall"
            >
              <h2>
                Latency
              </h2>
            </div>
            <hr
              class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginMedium"
            />
            <div
              id="latency"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
